CRM/Utils/System/Standalone.php

```diff
@@ -0,0 +1,690 @@
+<?php
+/*
+ +--------------------------------------------------------------------+
+ | Copyright CiviCRM LLC. All rights reserved.                        |
+ |                                                                    |
+ | This work is published under the GNU AGPLv3 license with some      |
+ | permitted exceptions and without any warranty. For full license    |
+ | and copyright information, see https://civicrm.org/licensing       |
+ +--------------------------------------------------------------------+
+ */
+
+/**
+ *
+ * @package CRM
+ * @copyright CiviCRM LLC https://civicrm.org/licensing
+ */
+
+/**
+ * Standalone specific stuff goes here.
+ */
+class CRM_Utils_System_Standalone extends CRM_Utils_System_Base {
+
+  /**
+   * @inheritdoc
+   */
+  public function initialize() {
+    parent::initialize();
+    $this->registerPathVars();
+  }
+
+  /**
+   * Specify the default computation for various paths/URLs.
+   */
+  protected function registerPathVars():void {
+    // Do nothing for now if being called from 'cv'
+    // @todo What about the command-line installer?
+    if (empty($_SERVER['SCRIPT_FILENAME']) || $_SERVER['SCRIPT_FILENAME'] == '/index.php') {
+      return;
+    }
+
+    Civi::paths()->register('civicrm.root', function () {
+      $webroot = dirname($_SERVER['SCRIPT_FILENAME']);
+      $root = dirname($webroot);
+      return [
+        'path' => implode(DIRECTORY_SEPARATOR, [$root, 'vendor', 'civicrm', 'civicrm-core']),
+        'url' => CIVICRM_UF_BASEURL . '/assets/civicrm/core',
+      ];
+    });
+    Civi::paths()->register('civicrm.files', function () {
+      $webroot = dirname($_SERVER['SCRIPT_FILENAME']);
+      $root = dirname($webroot);
+      return [
+        'path' => implode(DIRECTORY_SEPARATOR, [$root, 'web', 'upload']),
+        'url' => CIVICRM_UF_BASEURL . '/upload/',
+      ];
+    });
+    // For now, some places still assume they exist
+    $cmsRoot = function() {
+      $webroot = dirname($_SERVER['SCRIPT_FILENAME']);
+      return [
+        'path' => $webroot,
+        'url' => CIVICRM_UF_BASEURL . '/',
+      ];
+    };
+    Civi::paths()->register('cms', $cmsRoot);
+    Civi::paths()->register('cms.root', $cmsRoot);
+  }
+
+  /**
+   * @inheritdoc
+   */
+  public function getDefaultFileStorage() {
+    return [
+      'url' => 'upload',
+      // @todo Not sure if this is wise
+      'path' => $_SERVER['DOCUMENT_ROOT'],
+    ];
+  }
+
+  /**
+   * @inheritDoc
+   */
+  public function createUser(&$params, $mail) {
+    return FALSE;
+  }
+
+  /**
+   * @inheritDoc
+   */
+  public function updateCMSName($ufID, $email) {
+  }
+
+  /**
+   * @inheritDoc
+   */
+  public function getLoginURL($destination = '') {
+    $query = $destination ? ['destination' => $destination] : [];
+    return \Drupal\Core\Url::fromRoute('user.login', [], ['query' => $query])->toString();
```

Comment:
This comes up as a fatal error on an entityref field if you choose e.g. the New Individual popup. Maybe for now it could be `return CRM_Utils_System::url('civicrm/todo/user/login', 'reset=1');` or something like that, just so the popup works.

Reply:
@demeritcowboy Thanks - I updated to `civicrm/user/login`, since I think that makes sense as a login URL.